consensus/istanbul/backend/internal/replica/state.go

```diff
@@ -0,0 +1,342 @@
+				return err
+			}
+			rs.state = replicaPermanent
+			rs.startValidatingBlock = nil
+			rs.stopValidatingBlock = nil
+		}
+	case replicaWaiting:
+		if blockNumber.Cmp(rs.startValidatingBlock) >= 0 {
+			if err := rs.startFn(); err != nil {
+				logger.Warn("Error starting core", "err", err)
+				return err
+			}
+			if rs.stopValidatingBlock == nil {
+				logger.Info("Switching to primary (permanent)")
+				rs.state = primaryPermanent
+				rs.startValidatingBlock = nil
+				rs.stopValidatingBlock = nil
+			} else {
+				logger.Info("Switching to primary in range")
+				rs.state = primaryInRange
+			}
+		}
+	default:
+		// pass
+	}
+	return nil
+}
+
+// SetStartValidatingBlock sets the start block in the range [start, stop)
+func (rs *replicaStateImpl) SetStartValidatingBlock(blockNumber *big.Int) error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	defer rs.rsdb.StoreReplicaState(rs)
+
+	if blockNumber.Cmp(common.Big0) <= 0 {
+		return errors.New("blockNumber must be > 0")
+	}
+	if rs.stopValidatingBlock != nil && !(blockNumber.Cmp(rs.stopValidatingBlock) < 0) {
+		return errors.New("Start block number should be less than the stop block number")
+	}
+
+	switch rs.state {
+	case replicaPermanent:
+		rs.state = replicaWaiting
+	case replicaWaiting:
+		// pass. Changed start block while waiting to start.
+	default:
+		// todo error
+	}
+	rs.startValidatingBlock = blockNumber
+
+	return nil
+}
+
+// SetStopValidatingBlock sets the stop block in the range [start, stop)
+func (rs *replicaStateImpl) SetStopValidatingBlock(blockNumber *big.Int) error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	defer rs.rsdb.StoreReplicaState(rs)
+
+	if blockNumber.Cmp(common.Big0) <= 0 {
+		return errors.New("blockNumber must be > 0")
+	}
+	if rs.startValidatingBlock != nil && !(blockNumber.Cmp(rs.startValidatingBlock) > 0) {
+		return errors.New("Stop block number should be greater than the start block number")
+	}
+
+	switch rs.state {
+	case primaryPermanent:
+		rs.state = primaryInRange
+	case primaryInRange:
+		// pass. Changes stop block while waiting to stop.
+	case replicaPermanent:
+		// TODO: don't allow this / error
+		// Doesn't make sense, but not previously disallowed
+	case replicaWaiting:
+		// pass. Changed stop block while waiting to start.
+	}
+	rs.stopValidatingBlock = blockNumber
+
+	return nil
+}
+
+// MakeReplica makes this node a replica & clears all start/stop blocks.
+func (rs *replicaStateImpl) MakeReplica() error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	defer rs.rsdb.StoreReplicaState(rs)
+
+	if rs.state == primaryPermanent || rs.state == primaryInRange {
+		if err := rs.stopFn(); err != nil {
+			return err
+		}
+	}
+	rs.startValidatingBlock = nil
+	rs.stopValidatingBlock = nil
+	rs.state = replicaPermanent
+	return nil
+}
+
+// MakePrimary makes this node a primary & clears all start/stop blocks.
+func (rs *replicaStateImpl) MakePrimary() error {
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	defer rs.rsdb.StoreReplicaState(rs)
```

Comment:
what happens if this return error?

Reply:
even worse.... what happens if `startFn()` works but `StoreReplicaState` fails

Reply:
Probably want to revert the change if the save fails. It would be nice to be able to atomically do all of these state changes at once (change replica state & write to disk), but I'm not sure it's worth the added complexity.